osmosis-pbf2/src/main/java/org/openstreetmap/osmosis/pbf2/v0_6/PbfReader.java

```diff
@@ -39,34 +36,35 @@
 	 *            The number of worker threads for decoding PBF blocks.
 	 */
 	public PbfReader(File file, int workers) {
-		this.file = file;
-		this.workers = workers;
+		this(streamFor(file), workers);
```

Comment:
We need to instantiate "heavy" resources in the run method to ensure resources are cleaned up in failure scenarios.  If two classes instantiate resources in their constructors and the second one fails, the first one won't get cleaned up.  The run() methods trigger a full init/process/close sequence in downstream tasks which allows them to cleanup even if one of them fails to instantiate properly.

I realise it makes the code a bit funky with a number of optional code paths, but it's safer than the alternative.

To keep the code simpler, perhaps we could pass in an implementation of an InputStream factory interface that can be invoked in the run method.  That would allow the various ways of getting an InputStream to be separated from this class..=